include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -31,28 +31,63 @@ void instr_eval_execute(in instr_t instr, in MC_precomputed_t precomp, inout irr
 
 	if (is_bxdf && positiveCosFactor)
 	{
-		const float ior_scalar = colorToScalar(ior[0]);
-		float bxdf_eval_scalar_part;
-		uint ndf = instr_getNDF(instr);
-		float a = params_getAlpha(params);
-		float a2 = a*a;
-		float ay = params_getAlphaV(params);
-		float ay2 = ay*ay;
-		const vec3 albedo = params_getReflectance(params);
-
-		bool is_valid = true;
+		float eta;
+		float rcp_eta;
+
+		const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.isotropic.NdotV, 0.0);
+
+		bool is_valid = (NdotV > FLT_MIN);
 		bool refraction = false;
 #ifndef NO_BSDF
 		//here actually using stronger check for BSDF because it's probably worth it
-		if (op_isBSDF(op) && irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
+		if (is_bsdf)
 		{
-			is_valid = irr_glsl_calcAnisotropicMicrofacetCache(microfacet, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, ior_scalar, 1.0/ior_scalar);
-			refraction = true;
-		}
+			eta = colorToScalar(ior[0]);
+			rcp_eta = 1.0 / eta;
+			if (irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
+			{
+				is_valid = irr_glsl_calcAnisotropicMicrofacetCache(microfacet, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, eta, rcp_eta);
+				refraction = true;
+			}
+		} 
+		else
 #endif
+		{
+			microfacet = _microfacet;
+		}
 
 		if (is_valid)
 		{
+			float bxdf_eval_scalar_part;
+			const uint ndf = instr_getNDF(instr);
+			const float a = params_getAlpha(params);
+			const float a2 = a*a;
+#ifndef ALL_ISOTROPIC_BXDFS
+			const float ay = params_getAlphaV(params);
+			const float ay2 = ay*ay;
+#endif
+			const vec3 albedo = params_getReflectance(params);
+
+			const float NdotL = cosFactor;
+			const float NdotL2 = s.NdotL2;
+#ifndef ALL_ISOTROPIC_BXDFS
+			const float TdotL2 = s.TdotL * s.TdotL;
+			const float BdotL2 = s.BdotL * s.BdotL;
+#endif
+
+			const float NdotV2 = currInteraction.isotropic.NdotV_squared;
+#ifndef ALL_ISOTROPIC_BXDFS
+			const float TdotV2 = currInteraction.TdotV * currInteraction.TdotV;
+			const float BdotV2 = currInteraction.BdotV * currInteraction.BdotV;
+#endif
+
+			const float NdotH = irr_glsl_conditionalAbsOrMax(is_bsdf, microfacet.isotropic.NdotH, 0.0);
```

Comment:
no need to clamp `NdotH` by convention `H` must be in the positive hemisphere (otherwise `is_valid` will be false for bsdf and a brdf will not allow `L` and `V` outside of upper hemisphere, so half vector must also be in the hemisphere)